src/codegen/compile.ml

```diff
@@ -7462,12 +7467,40 @@ and compile_exp (env : E.t) ae exp =
     | ArrayPrim (m, t), es ->
       SR.Vanilla,
       Arr.lit env (List.map (compile_exp_vanilla env ae) es)
-    | IdxPrim, [e1; e2]  ->
+    | IdxPrim, [e1; e2] ->
       SR.Vanilla,
       compile_exp_vanilla env ae e1 ^^ (* offset to array *)
       compile_exp_vanilla env ae e2 ^^ (* idx *)
       Arr.idx_bigint env ^^
       load_ptr
+    | NextArrayOffset spacing, [e] ->
+      let advance_by =
+        match spacing with
+        | ElementSize -> Arr.element_size
+        | One -> 2l (* 1 : Nat *) in
+      SR.Vanilla,
+      compile_exp_vanilla env ae e ^^ (* previous byte offset to array *)
+      compile_add_const advance_by
+    | ValidArrayOffset, [e1; e2] ->
+      SR.bool,
+      compile_exp_vanilla env ae e1 ^^
+      compile_exp_vanilla env ae e2 ^^
+      G.i (Compare (Wasm.Values.I32 I32Op.LtU))
+    | DerefArrayOffset, [e1; e2] ->
+      SR.Vanilla,
+      compile_exp_vanilla env ae e1 ^^ (* skewed pointer to array *)
+      compile_exp_vanilla env ae e2 ^^ (* byte offset *)
+      G.i (Binary (Wasm.Values.I32 I32Op.Add)) ^^
+      Arr.(load_field 0l)              (* loads the element at the byte offset *)
```

Comment:
```suggestion
    | DerefArrayOffset, [e1; e2] ->
      SR.Vanilla,
      compile_exp_vanilla env ae e1 ^^ (* skewed pointer to array *)
      compile_exp_vanilla env ae e2 ^^ (* byte offset *)
      G.i (Binary (Wasm.Values.I32 I32Op.Add)) ^^
      Arr.load_field 0l            (* loads the element at the byte offset *)
```
Is there something clever going on here, because it seems odd. Is it safe to just add to the skewed pointer without unskewing first? And why return the 0th field of the resulting array on the stack?

Reply:
`Arr.load_field` implicitly unskews and hops over the administrative fields, then loads the element. Unskewing and hopping is done by an `add`, which commutes with the `add` of the gate offset. So yes, it is mildly clever :-)

Reply:
maybe just add a comment to explain this?